torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -99,6 +101,39 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
   }
 }
 
+IValue deserializeResptoIValueInternal(
+    RpcCommandBase& rpc,
+    MessageType messageType) {
+  switch (messageType) {
+    case MessageType::SCRIPT_RET: {
+      auto& ret = static_cast<ScriptResp&>(rpc);
+      return ret.value();
+    }
+    case MessageType::FORWARD_AUTOGRAD_RESP: {
+      auto& rpcWithAutograd = static_cast<autograd::RpcWithAutograd&>(rpc);
+
+      // Attach 'recv' autograd function.
+      addRecvRpcBackward(
+          rpcWithAutograd.autogradMetadata(),
+          rpcWithAutograd.tensors(),
+          rpcWithAutograd.fromWorkerId());
+
+      // Handle the original RPC.
+      auto wrappedMessageType = rpcWithAutograd.wrappedMessageType();
+      return deserializeResptoIValueInternal(
+          rpcWithAutograd.wrappedRpc(), wrappedMessageType);
```

Comment:
Can we move the autograd logic to `deserializeResponse`? That way python_functions.cpp and this method reuse the same logic instead of repeating it in both places?

Looks like `deserializeResponse` doesn't unwrap `FORWARD_AUTOGRAD_RESP`, we can change it to unwrap it and call `deserializeResponse` recursively. That way we share this logic in one place and simplifies our logic here to just checking whether the resp type is SCRIPT_RET.

Reply:
good idea!

Reply:
@pritamdamania87 I has a few tries, and can not get deserializeResponse() refactored cleanly if moving FORWARD_AUTOGRAD_RESP unwrap logic into deserializeResponse(), the reasons are following: 
1) FORWARD_AUTOGRAD_RESP message can only return RpcCommandBase& for wrappedRpc
2) deserializeResponse() needs to return wrappedMessageType, otherwise no way to know type of wrappedRpc

So we need interface like this: 

RpcCommandBase& deserializeResponse(const message& response, unique_ptr<RpcCommandBase>& rpc, MessageType& msgType)  // rpc unique_ptr is used to keep returned RpcCommandBase& alive

The users have to write codes like this:
unique_ptr<RpcCommandBase> rpc;
MessageType msgType = message.type();
auto& unwrappedRpc = deserializeResponse(message, rpc, msgType)
deserializeIValue(unwrappedRpc, msgType)

The above change does not sound clean to me. I prefer to keep deserializeResponse() simple as it is right now, how do you think?

Reply:
As discussed offline, lets move the `wrappedRpc_` out of RpcWithAutograd and add a messageType field to RpcCommandBase.